kernel/kernel-api/src/main/java/io/delta/kernel/internal/files/ParsedCheckpointData.java

```diff
@@ -65,4 +67,52 @@ protected ParsedCheckpointData(
     super(version, type, fileStatusOpt, inlineDataOpt);
     checkArgument(type.category == ParsedLogCategory.CHECKPOINT, "Must be a checkpoint");
   }
+
+  /** Here, returning 1 means that `this` is preferred over (i.e. greater than) `that`. */
+  @Override
+  public int compareTo(ParsedCheckpointData that) {
+    // Compare versions.
+    if (version != that.version) {
+      return Long.compare(version, that.version);
+    }
+
+    // Compare types.
+    if (type != that.type) {
+      return Integer.compare(type.ordinal(), that.type.ordinal());
+    }
+
+    // Use type-specific tiebreakers if versions and types are the same.
+    switch (type) {
+      case CLASSIC_CHECKPOINT:
+      case V2_CHECKPOINT:
+        return getTieBreaker(that);
+      case MULTIPART_CHECKPOINT:
+        final ParsedMultiPartCheckpointData thisCasted = (ParsedMultiPartCheckpointData) this;
+        final ParsedMultiPartCheckpointData thatCasted = (ParsedMultiPartCheckpointData) that;
+        final int numPartsComparison = Long.compare(thisCasted.numParts, thatCasted.numParts);
+        if (numPartsComparison != 0) {
+          return numPartsComparison;
+        } else {
+          return getTieBreaker(that);
+        }
```

Comment:
nit - this could be in ParsedMultiPartCheckpointData.compare? I'm not too picky tho